### PR TITLE
IMP: Allow for demuxing of reads with mixed orientation

### LINKED ISSUE #11: reorient/demux mixed-orientation reads
**Current Behavior**
- reads are in mixed orientation (both R1 and R2 reads contain forward/reverse reads)
- reads also often contain barcodes
- reads also probably contain primers

Sounds like this is the output produced by at least one company, so may want to provide support for this.

**Proposed Behavior**
1. add method to demux "mixed orientation" read format in one step
2. add a method to reorient reads (I think this may make more sense to allow a modular pipeline of import as paired-end --> reorient reads --> extract barcodes --> trim primers)

**References**
[raised on forum](https://forum.qiime2.org/t/problem-with-demux/2402?u=nicholas_bokulich)

### PR DESCRIPTION
**Pull request by Oddant1**
Closes #11. First iteration of allowing for demuxing of reads with multiple orientations by exposing a `mixed-orientation` flag that, when provided, reruns the demux with the orientations swapped. This is a basic implementation of what is discussed in the comments on issue #11 and, though I have tested it manually, this initial commit adds no new unit tests.

### CODE REVIEW

File: q2_cutadapt/_demux.py [hunk 1]
```diff
@@ -160,6 +160,15 @@ def _demux(seqs, forward_barcodes, reverse_barcodes, error_tolerance,
                                    current_untrimmed, error_tolerance,
                                    minimum_length)
         run_command(cmd)
+        if reverse_barcodes is not None and mixed_orientation:
+            temp = open_fhs['fwd']
```
Comment (thermokarst) [inline q2_cutadapt/_demux.py]: Prepare to have your mind blown :boom: 

```python
x = 4
y = 2
print(x, y)
# 4 2
```

```python
x, y = y, x
print(x, y)
# 2 4
```

File: q2_cutadapt/plugin_setup.py [hunk 2]
```diff
@@ -306,6 +307,10 @@
                           'the cutadapt default of 0 has been overridden, '
                           'because that value produces empty sequence '
                           'records.',
+        'mixed_orientation': 'Whether your data contains mixed orientation '
```
Comment (thermokarst) [inline q2_cutadapt/plugin_setup.py]: How about:
> Handle demultiplexing of mixed orientation reads (i.e. when forward and reverse reads coexist in the same file).

:stew:

File: q2_cutadapt/tests/test_demux.py [hunk 3]
```diff
@@ -294,6 +294,40 @@ def test_di_typical(self):
                          b'@id6\nTTTTTGCATGCA\n+\nzzzzzzzzzzzz\n']
         self.assert_untrimmed_results(exp_untrimmed, obs_untrimmed_art)
 
+    def test_multiple_orientations(self):
```
Comment (thermokarst) [inline q2_cutadapt/tests/test_demux.py]: Can you add another test that matches the schema defined here: https://cutadapt.readthedocs.io/en/stable/guide.html?demultiplexing-paired-end-reads-in-mixed-orientation

To clarify, the new test would not be dual barcoded, rather the barcode is either on the forward, or, on the reverse.

Comment (Oddant1) [reply q2_cutadapt/tests/test_demux.py]: I'm not sure I understand what you're asking for, that page talks about a lot of things, and the part about demuxing mixed orientations looks like what the test already does.

Comment (thermokarst) [reply q2_cutadapt/tests/test_demux.py]: There was a typo in the URL, check again.

Comment (thermokarst) [reply q2_cutadapt/tests/test_demux.py]: Please pay close attention to the "read layout" example in the cutadapt docs linked above - from what I can tell, this test does not look like it replicates the read layout example that the cutadapt authors have specified. It is important that we make sure we are able to handle the case where the reads are not dual barcoded. As a reminder, dual-barcoding is when a sample is identified by the combination of barcode in the forward and reverse read, as opposed to having a barcode only in the forward read or only in the reverse read.

Comment (Oddant1) [reply q2_cutadapt/tests/test_demux.py]: Ah ok, yeah this makes sense now. Thank you. So we want to add tests with adapters only at the beginning or only at the end of the read? Do we also want tests with adapters in the middle of the read?

**Comment by Oddant1** [comment]
I know there is a TODO in that last commit, it should hopefully be resolved in the next one

**Comment by Oddant1** [comment]
Pinging @thermokarst and @ebolyen because this came up on the [forum](https://forum.qiime2.org/t/cutadapt-demultiplexing-with-mixed-primers/15260/2) again the other day, and I think it was basically done? I don't remember why it stalled out.

**Comment by thermokarst** [comment]
>  I don't remember why it stalled out.

Nor I, but this is on my todo list.

**Comment by thermokarst** [comment]
> why did you decide to pass per_sample_sequences into _demux()?

Great question!

Previously, the helper `_demux` handled instantiating the directory format. For the mixed-orientation case though, I needed an easy way to write files into the directory format _more than once_. By injecting the instance in, I am now able to maintain the format's state outside of `_demux`.

Put another way, without that change, I would've wound up with two `per_sample_sequences` returned from my two calls to `_demux`. I would need to merge all the of the files in those two directories before returning from the Action, which is a huge pain. It is a pain because its not just as simple as copying the files into the same dir - you also have to concatenate any samples/files that are found in both `per_sample_sequences` directories.

> Looking good, fellas.

Thanks so much for looking things over!

**Pull request merged**